url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/alg/block/BlockInnerMultiplication.java

```diff
@@ -0,0 +1,562 @@
+        for( int i = 0; i < widthA; i++ , rowC += widthC) {
+            int colA = i + indexA;
+            int endA = colA + widthA*heightA;
+            int b = indexB;
+
+            // for( int k = 0; k < heightA; k++ ) {
+            while(colA != endA ) {
+                double valA = dataA[colA];
+
+                int c = rowC;
+                final int endB = b + widthC;
+
+                //for( int j = 0; j < widthC; j++ ) {
+                while( b != endB ) {
+                    dataC[ c++ ] -= valA * dataB[b++];
+                }
+                colA += widthA;
+            }
+        }
+    }
+
+    /**
+     * <p>
+     * Performs the follow operation on individual inner blocks:<br>
+     * <br>
+     * C = C - A * B<sup>T</sup>
+     * </p>
+     */
+    public static void blockMultMinusTransB( final double[] dataA, final double []dataB, final double []dataC,
+                                     int indexA, int indexB, int indexC,
+                                     final int heightA, final int widthA, final int widthC) {
+        for( int i = 0; i < heightA; i++ ) {
+            for( int j = 0; j < widthC; j++ ) {
+                double val = 0;
+
+                for( int k = 0; k < widthA; k++ ) {
+                    val += dataA[i*widthA + k + indexA] * dataB[j*widthA + k + indexB];
+                }
+
+                dataC[ i*widthC + j + indexC ] -= val;
+            }
+        }
+    }
+
+    /**
+     * <p>
+     * Performs the follow operation on individual inner blocks:<br>
+     * <br>
+     * C = A * B
+     * </p>
+     */
+    public static void blockMultSet( final double[] dataA, final double []dataB, final double []dataC,
+                                     int indexA, int indexB, int indexC,
+                                     final int heightA, final int widthA, final int widthC) {
+//        for( int i = 0; i < heightA; i++ ) {
+//            for( int k = 0; k < widthA; k++ ) {
+//                for( int j = 0; j < widthC; j++ ) {
+//                    dataC[ i*widthC + j + indexC ] += dataA[i*widthA + k + indexA] * dataB[k*widthC + j + indexB];
+//                }
+//            }
+//        }
+
+        int a = indexA;
+        int rowC = indexC;
+        for( int i = 0; i < heightA; i++ , rowC += widthC ) {
+            int b = indexB;
+
+            final int endC = rowC + widthC;
+            final int endA = a + widthA;
+            while( a != endA ) {//for( int k = 0; k < widthA; k++ ) {
+                double valA = dataA[a++];
+
+                int c = rowC;
+
+                if( b == indexB ) {
+                    while( c != endC  ) {//for( int j = 0; j < widthC; j++ ) {
+                        dataC[ c++ ] = valA * dataB[ b++ ];
+                    }
+                } else {
+                    while( c != endC  ) {//for( int j = 0; j < widthC; j++ ) {
+                        dataC[ c++ ] += valA * dataB[ b++ ];
+                    }
+                }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 266:274, 312:320, 480:488, 526:534.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.